traffic_ops/traffic_ops_golang/deliveryservice/acme.go

```diff
@@ -0,0 +1,363 @@
+	}
+	tx2.Commit()
+
+	api.CreateChangeLogRawTx(api.ApiChange, "DS: "+dsName+", ID: "+strconv.Itoa(*dsID)+", ACTION: Added SSL keys with "+acmeAccount.AcmeProvider, currentUser, logTx)
+
+	return nil
+}
+
+func getAcmeAccountConfig(cfg *config.Config, acmeProvider string) *config.ConfigAcmeAccount {
+	for _, acmeCfg := range cfg.AcmeAccounts {
+		if acmeCfg.AcmeProvider == acmeProvider {
+			return &acmeCfg
+		}
+	}
+	return nil
+}
+
+func getDSIdAndVersionFromName(db *sqlx.DB, xmlId string) (*int, *int64, error) {
+	var dsID int
+	var certVersion int64
+
+	if err := db.QueryRow(`SELECT id, ssl_key_version FROM deliveryservice WHERE xml_id = $1`, xmlId).Scan(&dsID, &certVersion); err != nil {
+		return nil, nil, err
+	}
+
+	return &dsID, &certVersion, nil
+}
+
+func GetAcmeClient(acmeAccount *config.ConfigAcmeAccount, userTx *sql.Tx, db *sqlx.DB) (*lego.Client, error) {
+	if acmeAccount.UserEmail == "" {
+		log.Errorf("An email address must be provided to use ACME with %v", acmeAccount.AcmeProvider)
+		return nil, errors.New("An email address must be provided to use ACME with " + acmeAccount.AcmeProvider)
+	}
+	storedAcmeInfo, err := getStoredAcmeAccountInfo(userTx, acmeAccount.UserEmail, acmeAccount.AcmeProvider)
+	if err != nil {
+		log.Errorf("Error finding stored ACME information: %s", err.Error())
+		return nil, err
+	}
+
+	myUser := MyUser{}
+	foundPreviousAccount := false
+	userPrivateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		log.Errorf("Error generating private key: %s", err.Error())
+		return nil, err
+	}
+
+	if storedAcmeInfo == nil || acmeAccount.UserEmail == "" {
+		myUser = MyUser{
+			key:   userPrivateKey,
+			Email: acmeAccount.UserEmail,
+		}
+	} else {
+		foundPreviousAccount = true
+		myUser = MyUser{
+			key:   &storedAcmeInfo.PrivateKey,
+			Email: storedAcmeInfo.Email,
+			Registration: &registration.Resource{
+				URI: storedAcmeInfo.URI,
+			},
+		}
+	}
+
+	config := lego.NewConfig(&myUser)
+	config.CADirURL = acmeAccount.AcmeUrl
+	config.Certificate.KeyType = certcrypto.RSA2048
+
+	client, err := lego.NewClient(config)
+	if err != nil {
+		log.Errorf("Error creating acme client: %s", err.Error())
+		return nil, err
+	}
+
+	if acmeAccount.AcmeProvider == tc.LetsEncryptAuthType {
+		client.Challenge.Remove(challenge.HTTP01)
+		client.Challenge.Remove(challenge.TLSALPN01)
+		trafficRouterDns := NewDNSProviderTrafficRouter()
+		trafficRouterDns.db = db
+		if err != nil {
+			log.Errorf("Error creating Traffic Router DNS provider: %s", err.Error())
+			return nil, err
+		}
+		client.Challenge.SetDNS01Provider(trafficRouterDns)
+	}
+
+	if foundPreviousAccount {
+		log.Debugf("Found existing account with %s", acmeAccount.AcmeProvider)
+		reg, err := client.Registration.QueryRegistration()
+		if err != nil {
+			log.Errorf("Error querying %s for existing account: %s", acmeAccount.AcmeProvider, err.Error())
+			return nil, err
+		}
+		myUser.Registration = reg
```

Comment:
Not sure if reg can be `nil` without an error?

Reply:
i dont think it can be nil without an error